### DISCUSSION TITLE
Bazel Remote Output Service

### DISCUSSION THREAD

**Discussion by coeuvre** [bazelbuild/bazel#20933] [Design proposals]
We propose to add a new flag `--experimental_remote_output_service` which accepts a URI pointing to a remote output service server that can manage Bazel's output tree (`bazel-out/`) with a userspace filesystem. We are also defining a gRPC protocol for the communication between Bazel and the service.

Please check the [design doc](https://docs.google.com/document/d/1W6Tqq8cndssnDI0yzFSoj95oezRKcIhU57nwLHaN1qk/edit) for details. All feedback is welcome!

Related to #12823.

**Comment by sluongng** [comment]
From reading the design doc, and the comments from reviewers, I wonder if this could be attached with a high-level overview on what are the goals we are trying to solve with this RPC. Then perhaps we could break the goals down into a multi-stage roadmap: What goals will be addressed in the first stage, what will be left for a later stage, and what will NOT be addressed (aka. non-goal)?

If we could break the RPC feature down categorically, then folks could rest assured that some of their needs could be addressed at a later stage. 🤔

**Comment by EdSchouten** [comment]
I just merged these changes on the Buildbarn side to use the new protocol:

- https://github.com/buildbarn/bb-remote-execution/commit/94b3776d5dbae7bc94c43ea70e5aa2e02f11f856
- https://github.com/buildbarn/bb-clientd/commit/ad659bdbe665ef66162263faec568f4e5ae25fe3

I have only been able to test this through unit tests. Not by actually using it. The reason obviously being that I don't have a copy of Bazel that implements the new protocol. I will patiently wait for @coeuvre to post his patch set for Bazel. I also invite others to give the latest version of bb_clientd a try by then.

**Comment by coeuvre** [comment]
A workable client is available in my branch https://github.com/coeuvre/bazel/tree/remote_output_service_head. I still need time to go through the review process and submit it but you can already try it now. Note that the feature is still pre-alpha so expect bugs and issues.

Thanks @EdSchouten for updating bb_clientd to the new protocol so I can use it as the server during the development.

To try it:
1. Use a `bazel` binary compiled from https://github.com/coeuvre/bazel/tree/remote_output_service_head
2. Setup [bb_clientd](https://github.com/buildbarn/bb-clientd)
3. Set `--remote_cache` or `--remote_executor` as usual.
4. Set `--experimental_remote_output_service` and `--experimental_remote_output_service_output_path_prefix`.

**Reply by scosenza** [reply]
Awesome work @coeuvre! Can you or @EdSchouten provide more explicit instructions for what values to use for the new experimental flags? 
I tried to guess from the the bb_clientd docs, but this failed:

Note that bazelros is the custom compiled bazel from [remote_output_service_head](https://github.com/coeuvre/bazel/tree/remote_output_service_head).

```
steve_cosenza@steve-cosenzas-MacBook-Pro treehouse % lsof /Users/steve_cosenza/.cache/bb_clientd/grpc
COMMAND     PID          USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
bb_client 48535 steve_cosenza   19u  unix 0x619a995d0c92f5cb      0t0      /Users/steve_cosenza/.cache/bb_clientd/grpc

steve_cosenza@steve-cosenzas-MacBook-Pro treehouse % ls /Users/steve_cosenza/bb_clientd/cas/engflow/blobs/sha256
command		directory	executable	file		tree

steve_cosenza@steve-cosenzas-MacBook-Pro treehouse % bazelros build //projects/banana/... --experimental_remote_output_service=unix:/Users/steve_cosenza/.cache/bb_clientd/grpc --experimental_remote_output_service_output_path_prefix=/Users/steve_cosenza/bb_clientd/cas/engflow/blobs/sha256/file/
Starting local Bazel server and connecting to it...
INFO: Invocation ID: 4d8f53c5-6f56-4002-944f-bb5fb5551c4d
Loading: 1 packages loaded
FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.NullPointerException
	at com.google.devtools.build.lib.remote.BazelOutputServiceREv2Proto$StartBuildArgs$Builder.setRemoteCache(BazelOutputServiceREv2Proto.java:700)
	at com.google.devtools.build.lib.remote.BazelOutputService.startBuild(BazelOutputService.java:155)
```

Thanks in advance for any help!

**Reply by coeuvre** [reply]
Depending how you setup the server. In case of bb_clientd, I used the default config. So the values are `--experimental_remote_output_service=unix:$HOME/.cache/bb_clientd/grpc --experimental_remote_output_service_output_path_prefix=$HOME/bb_clientd/outputs`.

**Reply by EdSchouten** [reply]
Indeed. Please use the `/outputs/` directory as the output path prefix. Don't use `/cas/*`. The former is a special purpose writable directory that Bazel can use to construct `bazel-out/` directories. The path you use is simply a read-only space that can be used to explore raw contents of the CAS.

**Reply by coeuvre** [reply]
In addition to the two new experimental flags, you also need to set `--remote_cache` or `--remote_executor`. I will update the code to print the hint out instead of crashing Bazel.

**Reply by EdSchouten** [reply]
Also: I see you're using bb_clientd on macOS. On macOS we need to use NFSv4.0, which lacks some facilities in the area of cache invalidation which we need for Remote Output Service to work reliably/correctly. NFSv4.2 would solve this, but this is not supported by the macOS NFS driver.

So on macOS, bb_clientd is mainly just usable as a local cache, or to explore CAS contents through `~/bb_clientd/cas/*`. For Remote Output Service you'd better use Linux for time being...

**Reply by scosenza** [reply]
@coeuvre The custom compiled bazel wasn't triggering our normal wrapper which adds the RBE flags. Thanks for the pointer!
@EdSchouten Ah, that's unfortunate :-( I recall you saying the Sonoma 14.4 improves NFS. Is it still missing the invalidation features?

**Reply by EdSchouten** [reply]
Yes, unfortunately it still does. Sonoma 14.4 includes many improvements to the NFS client driver that are of use on the worker side. It makes `bb_worker`'s virtual output directory work (almost) flawlessly. However, it still only implements NFSv4.0, which isn't sufficient.

For Remote Output Service to work well using NFS, we need:

- NFSv4.1/4.2's support for parallel `OPEN` calls by a single open-owner. With NFSv4.0, all `OPEN` calls issued against the NFS mount are essentially serialised, which is very bad for performance.
- NFSv4.1/4.2's support for sending callbacks in case directories are modified (`CB_NOTIFY`). Otherwise macOS won't be notified when `StageArtifacts()` mutates the file system behind the scenes.

**Reply by scosenza** [reply]
@EdSchouten For some background (which I'm sure you've heard many times before), our engineers all have Mac laptops, so I was looking into lazy file loading to avoid eagerly downloading files over home internet connections. And while ROS can also help our builds from Linux clients, our Linux machines are already in the datacenter, so network speeds are much less of an issue there. Either way, thanks again for all of your work in this area!

**Reply by coeuvre** [reply]
> * NFSv4.1/4.2's support for sending callbacks in case directories are modified (`CB_NOTIFY`). Otherwise macOS won't be notified when `StageArtifacts()` mutates the file system behind the scenes.

I assume this is for `StartBuildResponse.initial_output_path_contents`? For NFSv4.0, can we simply don't track the modifications and let Bazel figure out which output file has been modified? This might be slower but the overall UX should be better than not using remote output service at all.

**Reply by EdSchouten** [reply]
@coeuvre It's unrelated to that feature. Think of it like this:

1. Bazel `stat()`s a file in `bazel-out/` and notices it's absent.
2. Bazel kicks off a remote action to get that file built.
3. Bazel will call `StageArtifacts()` to create that file in `bazel-out/`.

If Bazel (or a locally running build action) would attempt to `stat()` that file not long afterwards, they might still get a file-not-found, because the kernel caches the fact that the file does not exist. Or even worse, if the file did exist but it had to be rebuilt, `stat()` would see the old file. This is because the file system's contents changed behind the kernel's back.

Right now we solve that in bb_clientd by setting mount options `acdirmin=0,acdirmax=0` to set caching of directory contents to zero seconds, but this is obviously _very_ bad for performance. And I think that even in that case it doesn't rule out some false caching. For it to work reliably, we need a mechanism where bb_clientd can tell the kernel to purge the name cache entry.

**Reply by moroten** [reply]
A build of the `release-7.2.0` branch, as of 2024-04-16, with cherry-picks from [remote_output_service_head](https://github.com/coeuvre/bazel/tree/remote_output_service_head) is available at https://github.com/meroton/bazel/releases/tag/7.2.0-pre20240416ros.

**Comment by coeuvre** [comment]
All changes have been cherry-picked to 7.2 branch. You can try it out with the upcoming 7.2 release!

**Comment by adincebic** [comment]
@buchgr is this on the roadmap for the bazel-remote?

**Reply by brentleyjones** [reply]
@mostynb is the maintainer of bazel-remote now.

**Reply by mostynb** [reply]
This is not currently on bazel-remote's roadmap. But it might be possible for other implementations of the fuse service to use bazel-remote as the cache service.

**Reply by EdSchouten** [reply]
Exactly. It should be possible to use bb_clientd in combination with bazel-remote.
